src/hotspot/share/memory/allocation.hpp

```diff
@@ -216,6 +216,48 @@ template <MEMFLAGS F> class CHeapObj {
   void  operator delete [] (void* p) { FreeHeap(p); }
 };
 
+class DynCHeapObj {
```

Comment:
I tried this and it seems to work:  If you make CHeapObj new operators take MEMFLAGS f = F, where F is the default  value, you could not have this class and make outputStream inherit from CHeapObj<mtInternal> and the new calls will override mtInternal.  I guess the risk is that you get mtInternal if you forget the parameter to new.

Reply:
Good idea! I implemented this instead.

Reply:
We can eliminate this problem:
"I guess the risk is that you get mtInternal if you forget the parameter to new."

We can instead have CHeapObj<mtFoo> that *only* allocates with mtFoo, and
CHeapObj<> that *requires* specifying the flag when allocating the memory. (I
did something like this for the CHeapAllocator in my prototype for Standard
Library usage.) As part of that, we can also have the MEMFLAGS argument (when
needed) always be the first argument in the placement list.

Very(!) lightly tested:
https://github.com/kimbarrett/openjdk-jdk/tree/dynamic-memflags

Reply:
Or you could have the class that requires dynamic mtFlag be declared with mtNone and assert in one place in AllocateHeap that mt != mtNone.  Actually that assert might already be there.  Can't really puzzle out what Kim's change does.

Reply:
I added a mtInvalid flag and did the assert. mtNone seems to be dealt with in some code, I'm not sure if it's appropriate to use it as an invalid value. Regardless, mtInvalid is a better name (for this purpose), so maybe we can change mtNone to mtInvalid in the future?

Reply:
A much simpler (in the sense of avoiding the use of advanced template features) alternative to the one I posted earlier.  (Thanks to @stefank for the suggestion.)
https://github.com/openjdk/jdk/compare/master...kimbarrett:openjdk-jdk:dynamic-memflags-specialize
I dislike the current proposed approach of adding an optional MEMFLAGS argument to all the CHeapObj allocation functions.